cmd/influxd/upgrade/security_test.go

```diff
@@ -0,0 +1,248 @@
+package upgrade
+
+import (
+	"bufio"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"sort"
+	"strings"
+	"testing"
+	"unsafe"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/influxdb/v2/v1/services/meta"
+	"github.com/influxdata/influxql"
+	"go.uber.org/zap"
+)
+
+func TestGenerateScript(t *testing.T) {
+
+	type testCase struct {
+		name string
+		users []meta.UserInfo
+		db2ids map[string][]string
+		want string
+		wantErr error
+	}
+
+	var testCases = []testCase{
+		{
+			name: "ordinary",
+			users: []meta.UserInfo{
+				{ // not upgraded because admin
+					Name:  "superman",
+					Admin: true,
+				},
+				{ // not upgraded because no privileges
+					Name:  "loser",
+					Admin: false,
+				},
+				{
+					Name:  "weatherman",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"water": influxql.AllPrivileges,
+						"air": influxql.AllPrivileges,
+					},
+				},
+				{
+					Name:  "hitgirl",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"hits": influxql.WritePrivilege,
+					},
+				},
+				{
+					Name:  "boss@hits.org", // special name
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"hits": influxql.AllPrivileges,
+					},
+				},
+				{
+					Name:  "viewer",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"water": influxql.ReadPrivilege,
+						"air": influxql.ReadPrivilege,
+					},
+				},
+			},
+			db2ids: map[string][]string{
+				"water": {"33f9d67bc9cbc5b7", "33f9d67bc9cbc5b8", "33f9d67bc9cbc5b9"},
+				"air": {"43f9d67bc9cbc5b7", "43f9d67bc9cbc5b8", "43f9d67bc9cbc5b9"},
+				"hits": {"53f9d67bc9cbc5b7"},
+			},
+			want: testScriptShOrdinary,
+		},
+		{
+			name: "missing buckets",
+			users: []meta.UserInfo{
+				{
+					Name:  "weatherman",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"water": influxql.AllPrivileges,
+						"air": influxql.AllPrivileges,
+					},
+				},
+			},
+			db2ids: nil,
+			wantErr: errors.New("upgrade: there were errors/warnings, please fix them and run the command again"),
+		},
+		{
+			name: "no users",
+			users: []meta.UserInfo{},
+			want: "",
+		},
+	}
+
+	var shSuffix string
+	if runtime.GOOS == "win" {
+		shSuffix = ".cmd"
+	} else {
+		shSuffix = ".sh"
+	}
+	shSuffix = ".cmd"
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) { // better do not run in parallel
+			// mock v1 meta
+			v1 := &influxDBv1{
+				meta: &meta.Client{},
+			}
+			data := &meta.Data{
+				Users: tc.users,
+			}
+			// inject users into mock v1 meta client
+			f := reflect.ValueOf(v1.meta).Elem().Field(4)
+			f = reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr())).Elem()
+			f.Set(reflect.ValueOf(data))
+
+			// script output file
+			tmpfile, err := ioutil.TempFile("", "upgrade-security-*" + shSuffix)
+			if err != nil {
+				t.Fatal(err)
+			}
+			tmpfile.Close()
```

Comment:
Go 1.15 now has support for temporary directories that are automatically cleaned up on completion of the test:

https://golang.org/doc/go1.15#testing

```go
tmpFile, err := ioutil.TempFile(t.TempDir(), "upgrade-security-*" + shSuffix)
require.NoError(err)
require.NoError(tmpfile.Close())
```

Reply:
Thank you! I did not know this, I've also missed the `go` directive change to 1.15 in `go.mod`.